### PR TITLE
[TFLite] QNN support for TFLite 2.1.0 quantized models

### PR DESCRIPTION
**Pull request by anijain2305**
TFLite 2.1.0 has changed the quantization specs from older TFLite versions. The specs are present at - https://www.tensorflow.org/lite/performance/quantization_spec

### Key changes
* No TFLite2.0 hosted models. Need a way to create quantized models. Dependency on tensorflow_hub package (@tqchen 
 please advise here).
* Per-axis (aka channel) quantization
* Int8 datatypes
* Extra restrictions per op regarding scale and zero points
* Quantize TFLite op also works as Requantize

### CODE REVIEW

**Comment by anijain2305** [comment]
Opening the PR for review. @siju-samuel @masahi @u99127  @tqchen 

The CI fails because of a missing package - tensorflow_hub

File: python/tvm/relay/frontend/tflite.py [hunk 1]
```diff
@@ -2566,17 +2613,27 @@ def convert_quantize(self, op):
...
+        if input_tensor_type_str == "float32":
+            out = self.quantize(in_expr, output_tensor)
+        else:
+            out = _qnn.op.requantize(in_expr,
+                                     input_scale=input_tensor.qnn_params['scale'],
+                                     input_zero_point=input_tensor.qnn_params['zero_point'],
+                                     output_scale=output_tensor.qnn_params['scale'],
+                                     output_zero_point=output_tensor.qnn_params['zero_point'],
+                                     out_dtype=output_tensor_type_str)
         return out
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:2637]: You are right. I will add a test case in this PR. This will enable us to keep those 5 end to end tests as well.

**Comment by anijain2305** [comment]
Thanks @u99127 for the review. All make sense.  I will add clarifications.

File: python/tvm/relay/frontend/tflite.py [hunk 2]
```diff
@@ -262,21 +298,25 @@ def get_tensor_value(self, tensor_wrapper):
...
+
+        if tensor_wrapper.tensor.Type() == TensorType.INT8:
+            return np.frombuffer(data, dtype=np.int8).reshape(shape)
         if tensor_wrapper.tensor.Type() == TensorType.UINT8:
-            return np.frombuffer(tensor_wrapper.buffer.DataAsNumpy(), dtype=np.uint8).reshape(
-                tensor_wrapper.tensor.ShapeAsNumpy())
-        if tensor_wrapper.tensor.Type() == TensorType.FLOAT32:
-            return np.frombuffer(tensor_wrapper.buffer.DataAsNumpy(), dtype=np.float32).reshape(
-                tensor_wrapper.tensor.ShapeAsNumpy())
+            return np.frombuffer(data, dtype=np.uint8).reshape(shape)
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:311]: Can we add support for int16/float16 as well?

File: python/tvm/relay/frontend/tflite.py [hunk 3]
```diff
@@ -1605,7 +1645,7 @@ def convert_fully_connected(self, op):
 
         # weight tensor type should be UINT8 (quantization) or FLOAT32
         weight_tensor_type = weight_tensor.tensor.Type()
-        assert weight_tensor_type in (TensorType.UINT8, TensorType.FLOAT32)
+        assert weight_tensor_type in (TensorType.INT8, TensorType.UINT8, TensorType.FLOAT32)
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:1648]: change the comment in line 1646 as well to include INT8. check similar in other places also

File: python/tvm/relay/frontend/tflite.py [hunk 4]
```diff
@@ -2723,6 +2779,15 @@ def get_scalar_from_constant(expr):
         "value must be float32/int32"
     return np.asscalar(value)
 
+def get_vector_from_constant(expr):
+    """ Returns scalar value from Relay constant scalar. """
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:2783]: Update docstring

File: tests/python/frontend/tflite/test_forward.py [hunk 5]
```diff
@@ -2445,6 +2467,112 @@ def test_forward_qnn_mobilenet_v3_net():
...
+        tflite_model_buf = _quantize_tf_hub_keras_model(url, 224, 224)
+        data = pre_processed_image(224, 224)
+
+        tflite_output = run_tflite_graph(tflite_model_buf, data)
+        tflite_predictions = np.squeeze(tflite_output)
+        tflite_sorted_labels = tflite_predictions.argsort()[-3:][::-1]
+        tvm_output = run_tvm_graph(tflite_model_buf, np.array(data), 'input_1')
+        tvm_predictions = np.squeeze(tvm_output)
+        tvm_sorted_labels = tvm_predictions.argsort()[-3:][::-1]
+        tvm.testing.assert_allclose(tvm_sorted_labels, tflite_sorted_labels)
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:2505]: just for info, How different is the output values? if comparable, we can do the output value comparison itself. Since rounding method is different in tflite and tvm and due to that there are some difference in  outputs.

File: tests/python/frontend/tflite/test_forward.py [hunk 6]
```diff
@@ -2445,6 +2467,112 @@ def test_forward_qnn_mobilenet_v3_net():
     tvm.testing.assert_allclose(tvm_sorted_labels, tflite_sorted_labels)
 
 
+
+def _quantize_tf_hub_keras_model(url, height, width):
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:2471]: Add func headers similar to previous testcases.

File: tests/python/frontend/tflite/test_forward.py [hunk 7]
```diff
@@ -73,6 +74,28 @@ def get_real_image(im_height, im_width):
...
+        if image.dtype != tf.float32:
+            image = tf.image.convert_image_dtype(image, dtype=tf.float32)
+        image = tf.image.central_crop(image, central_fraction=0.875)
+    # Resize the image to the specified height and width.
+    image = tf.expand_dims(image, 0)
+    image = tf.image.resize(image, [height, width],
+                            align_corners=False)
+    image = tf.image.resize(image, [height, width])
+    image = tf.squeeze(image, [0])
+    image = tf.expand_dims(image, axis=0)
```
Comment (siju-samuel) [inline tests/python/frontend/tflite/test_forward.py:95]: i think this squeeze and expanddims are no reqiuired since both are done at axis 0.

File: python/tvm/relay/frontend/tflite.py (hunk 2 above)
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:311]: Samuel. Thats a good point. I looked into int16 support. This is experimental in TFLite for now. 

For float16, I think we will need a good number of tests. I think we can handle float16 in a separate PR.

File: python/tvm/relay/frontend/tflite.py (hunk 3 above)
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:1648]: Good catch!

File: tests/python/frontend/tflite/test_forward.py (hunk 5 above)
Comment (anijain2305) [reply tests/python/frontend/tflite/test_forward.py:2505]: Yes, rounding differences lead to these differences (although minimal effect on Top1/5 accuracy). I tried softmax'd output for older TFLite 1.13 tests and labels looked better for testing. I will check the TFLite 2.1.0 tests (I dont think picture would change)

File: python/tvm/relay/frontend/tflite.py [hunk 8]
```diff
@@ -262,21 +298,25 @@ def get_tensor_value(self, tensor_wrapper):
         except ImportError:
             raise ImportError("The tflite package must be installed")
 
+        data = tensor_wrapper.buffer.DataAsNumpy()
+        shape = tensor_wrapper.tensor.ShapeAsNumpy()
+
+        # Set shape to 1 if the data is a scalar type
+        if data.shape == (1,) and isinstance(shape, int) and shape == 0:
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py]: I modified the comments. Can you take a look now? This is sort of a corner case when the TFLite buffer is a scalar.

File: python/tvm/relay/frontend/tflite.py [hunk 9]
```diff
@@ -298,11 +303,15 @@ def get_tensor_value(self, tensor_wrapper):
...
+        # Read the data from the buffer. Also extract the shape.
+        # The shape is used later to reshape the data.
         data = tensor_wrapper.buffer.DataAsNumpy()
         shape = tensor_wrapper.tensor.ShapeAsNumpy()
 
-        # Set shape to 1 if the data is a scalar type
-        if data.shape == (1,) and isinstance(shape, int) and shape == 0:
+        # When TFLite buffer is of size 1 (scalar), then TFLite tensor shape is set to 0.
+        # Therefore, we set the shape to 1 for numpy reshape to work.
+        Set shape to 1 if the data is a scalar type
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:313]: missed to add # before comment?

File: python/tvm/relay/frontend/tflite.py [hunk 10]
```diff
@@ -258,15 +260,18 @@ def get_tensors(self, tensors_idx_list):
...
+                            raise tvm.error.OpAttributeInvalid(\
+                                    "TFLite per-axis quantization restricts all zero points to be"
+                                    + " 0, but a non-zero value is observed")
                         zero_point = int(zero_point[0])
 
                     # Scalar - Per-tensor quantization
-                    elif tflite_scale.shape == (1,) and tflite_zero_point.shape == (1,):
+                    elif tflite_scale.size == 1 and tflite_zero_point.size == 1:
                         scale = float(tflite_scale[0])
                         zero_point = int(tflite_zero_point[0])
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:277]: error coud be due to wrong `tflite_zero_point`as well,  plz print type of `tflite_zero_point` in else case.

File: python/tvm/relay/frontend/tflite.py [hunk 11]
```diff
@@ -258,15 +260,18 @@ def get_tensors(self, tensors_idx_list):
...
 
                     # Tensor - Per-axis quantization
-                    if tflite_scale.shape != (1,) and tflite_zero_point.shape != (1,):
+                    if tflite_scale.size != 1 and tflite_zero_point.size != 1:
                         scale = tflite_scale
-                        # Ensure that all zero points are identical
+                        # Ensure that all zero points are zeros
                         zero_point = tflite_zero_point
-                        assert all(x == zero_point[0] for x in zero_point)
+                        if not all(x == 0 for x in zero_point):
```
Comment (siju-samuel) [inline python/tvm/relay/frontend/tflite.py:267]: Suggestion `if not all(x == 0 for x in zero_point):` >> `if not np.all((zero_point == 0)):`

File: python/tvm/relay/frontend/tflite.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:2679]: image

Above test case added to force both types of quantize nodes

**Comment by anijain2305** [comment]
@u99127 @siju-samuel CI passes now. Can you please take another look? There were more changes, mostly to add unit level tests.

**Comment by siju-samuel** [comment]
LGTM, Any more changes pending?

**Comment by anijain2305** [comment]
@u99127 ping

File: python/tvm/relay/frontend/tflite.py [hunk 12]
```diff
@@ -263,21 +305,29 @@ def get_tensor_value(self, tensor_wrapper):
...
+        data = tensor_wrapper.buffer.DataAsNumpy()
+        shape = tensor_wrapper.tensor.ShapeAsNumpy()
+
+        # When TFLite buffer is of size 1 (scalar), then TFLite tensor shape is set to 0.
+        # Therefore, we set the shape to 1 for numpy reshape to work.  Set shape to 1 if the data is
+        # a scalar type
+        if data.size == 1 and isinstance(shape, int) and shape == 0:
+            shape = (1,)
+
+        if tensor_wrapper.tensor.Type() == TensorType.INT8:
```
Comment (u99127) [inline python/tvm/relay/frontend/tflite.py:319]: Minor nit and this should really be credited to Dmitriy Smirnov. https://github.com/d-smirnov

the condition here could well be pulled out into a helper function that has a dictionary to help us map from TensorType to numpy type ? 

Would make the code much cleaner and reduce duplication. 

i.e. something like 


def get_tensor_type_as_numpy(self, tensor_wrapper):
        """Returns np.dtype out of TensorType"""
        """Returns np.dtype out of TensorType"""
        assert isinstance(tensor_wrapper, TensorWrapper)

        try:
            from tflite.TensorType import TensorType
            return { TensorType.UINT8:   np.uint8,
                     TensorType.INT8:    np.int8,   
                     TensorType.FLOAT32: np.float32,
                     TensorType.INT32:   np.int32,
                     TensorType.INT64:   np.int64,
                     TensorType.BOOL:    np.bool_ }[ tensor_wrapper.tensor.Type() ]
        except ImportError:
            raise ImportError("The tflite package must be installed")

        except KeyError:
            raise NotImplementedError("Tensor type '{}' currently not supported"
                                      .format(tensor_wrapper.tensor.Type()))

    def get_tensor_value(self, tensor_wrapper):
        """Get tensor buffer value from given tensor wrapper"""
        assert isinstance(tensor_wrapper, TensorWrapper)

        value_type = self.get_tensor_type_as_numpy( tensor_wrapper )
        return np.frombuffer( tensor_wrapper.buffer.DataAsNumpy(), dtype=value_type ).reshape(
                tensor_wrapper.tensor.ShapeAsNumpy() \
                    if 0 != tensor_wrapper.tensor.ShapeLength() \
                        else [] )

File: python/tvm/relay/frontend/tflite.py [hunk 13]
```diff
@@ -651,12 +701,43 @@ def convert_shape(self, op):
 
     def convert_relu(self, op):
         """Convert TFLite ReLU"""
+        try:
+            from tflite.ActivationFunctionType import ActivationFunctionType
+        except ImportError:
+            raise ImportError("The tflite package must be installed")
+
```
Comment (u99127) [inline python/tvm/relay/frontend/tflite.py:708]: I think this is unnecessary given the import of ActivationFunctionType in the constructor [here](https://github.com/apache/incubator-tvm/blob/b979bf6a7630ada055fff2d65e1cd0f8d55bb6a0/python/tvm/relay/frontend/tflite.py#L54)

File: python/tvm/relay/frontend/tflite.py [hunk 14]
```diff
@@ -692,6 +773,11 @@ def _hard_swish(data):
 
     def convert_relu6(self, op):
         """Convert TFLite ReLU6"""
+        try:
+            from tflite.ActivationFunctionType import ActivationFunctionType
```
Comment (u99127) [inline python/tvm/relay/frontend/tflite.py:777]: Same as relu,  I think this is unnecessary given the import of ActivationFunctionType in the constructor [here](https://github.com/apache/incubator-tvm/blob/b979bf6a7630ada055fff2d65e1cd0f8d55bb6a0/python/tvm/relay/frontend/tflite.py#L54)

File: tests/python/frontend/tflite/test_forward.py [hunk 15]
```diff
@@ -787,14 +897,30 @@ def test_forward_convolution():
...
-    _test_convolution([4, 17, 17, 12], [3, 3, 12, 2], [1, 1], [2, 2], 'VALID', 'NHWC', True)
-    # dephtwise convolution with single input channel
-    _test_convolution([1, 76, 64, 1], [9, 5, 1, 96], [1, 1], [1, 1], 'SAME', 'NHWC', True)
+        # depthwise convolution
+        _test_convolution([4, 8, 8, 176], [1, 1, 176, 1], [1, 1], [1, 1], 'SAME', 'NHWC', True, quantized=quantized)
+        _test_convolution([4, 17, 17, 19], [3, 3, 19, 1], [1, 1], [2, 2], 'VALID', 'NHWC', True, quantized=quantized)
+        _test_convolution([4, 17, 17, 124], [1, 1, 124, 1], [1, 1], [1, 1], 'SAME', 'NHWC', True, quantized=quantized)
+        _test_convolution([4, 17, 17, 12], [3, 3, 12, 1], [1, 1], [2, 2], 'VALID', 'NHWC', True, quantized=quantized)
+        _test_convolution([4, 17, 17, 12], [3, 3, 12, 2], [1, 1], [2, 2], 'VALID', 'NHWC', True, quantized=quantized)
+        # dephtwise convolution with single input channel
```
Comment (u99127) [inline tests/python/frontend/tflite/test_forward.py:906]: dephtwise / depthwise.

File: python/tvm/relay/frontend/tflite.py (hunk 12 above)
Comment (d-smirnov) [reply python/tvm/relay/frontend/tflite.py:319]: Please consider following change as well:
```
    def has_same_qnn_params(self, lhs_tensor, rhs_tensor):
        lhs_scale = lhs_tensor.qnn_params['scale']
        rhs_scale = rhs_tensor.qnn_params['scale']
        lhs_zero_point = lhs_tensor.qnn_params['zero_point']
        rhs_zero_point = rhs_tensor.qnn_params['zero_point']
        return np.allclose( lhs_scale.data.asnumpy(), rhs_scale.data.asnumpy(), rtol=1e-5, atol=1e-5 ) and \
               np.allclose( lhs_zero_point.data.asnumpy(), rhs_zero_point.data.asnumpy(), rtol=1e-5, atol=1e-5 )


```

Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:319]: For the first comment, thanks, let me take a look.

For the second suggestion for has_same_qnn_params, I think we do not need that. For all the ops where we have to check params are same, they have scalar scale and zero point. This is because per-axis quantization is limited to weights, and thus limited to conv2d and dense op where we do not need this check.

File: python/tvm/relay/frontend/tflite.py (hunk 13 above)
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:708]: Tried this but it failed, the scope of imports is limited to the functions in which they are imported.

File: python/tvm/relay/frontend/tflite.py (hunk 14 above)
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py:777]: Same as before

**Comment by anijain2305** [comment]
@u99127 can you please take a look again?

**Comment by u99127** [comment]
> 
> 
> @u99127 can you please take a look again?

LGTM.

**Comment by anijain2305** [comment]
@siju-samuel Please approve and merge if its looks good :)

**Comment by siju-samuel** [comment]
Thanks @anijain2305 @u99127 @tqchen. This is now merged.
